### PR TITLE
Consolidate and simplify kotlinx-metadata interop

### PR DESCRIPTION
**Pull request by ZacSweers**
This rehashes a previous attempt to simplify the kotlinx-metadata artifacts into a single artifact.

The commits here are intended to be focused and easily digestible.

The final structure is we now have a single `interop:kotlinx-metadata` project and analogous single artifact. The previous artifacts (core, specs, classinspectors) are all merged into this project now. The previous `specs-tests` are also now merged into this project's tests.

API changes:
- Immutable implementations of Km types are now removed. They didn't offer much value, their equality/hashcode semantics were unstable due to the underlying types not always implementing them, and in general just unnecessary. Everything runs through the main `Km` types
- One change as a result of the previous bullet is we had to implement a custom comparator for Maps with `Km` types, since the raw ones didn't implement equals. The custom comparators just use names, signatures, and (as a last resort) a simple computed key.

Most of the previous public APIs are the same otherwise! Considering most usages just converted to the immutable type and then to the typespec, adoption of this should be fairly simple.

### CODE REVIEW

File: docs/interop-kotlinx-metadata.md [hunk 1]
```diff
@@ -35,21 +35,50 @@ println("Is class? ${kmClass.isClass}")
...
+represented sources can be incomplete without this information available. Reflective and javax
+`Elements` implementations are available under the `kotlinpoet-classinspector-*` artifacts.
+
+Generated sources are solely _stub_ implementations, meaning implementation details of elements
+like functions, property getters, and delegated properties are simply stubbed with `TODO()`
+placeholders.
+
+### Known limitations
+
+- Only `KotlinClassMetadata.Class` supported for now. No support for `FileFacade`, `SyntheticClass`, `MultiFileClassFacade`, or `MultiFileClassPart`
```
Comment (ZacSweers) [inline docs/interop-kotlinx-metadata.md:76]:
```suggestion
- Only `KotlinClassMetadata.Class` and `KotlinClassMetadata.FileFacade` are supported for now. No support for `SyntheticClass`, `MultiFileClassFacade`, or `MultiFileClassPart`
```

Out of date

File: docs/interop-kotlinx-metadata.md [hunk 2]
```diff
@@ -35,21 +35,50 @@ println("Is class? ${kmClass.isClass}")
...
+- `@file:` annotations are not supported yet.
```
Comment (ZacSweers) [inline docs/interop-kotlinx-metadata.md:77]:
```suggestion
```

No longer true!

File: docs/interop-kotlinx-metadata.md [hunk 3]
```diff
@@ -35,21 +35,50 @@ println("Is class? ${kmClass.isClass}")
...
+- `@JvmOverloads` annotations are only supported with `ElementsClassInspector` and not reflection.
+- Non-const literal values are only supported with `ReflectiveClassInspector` and not reflection.
```
Comment (ZacSweers) [inline docs/interop-kotlinx-metadata.md:79]:
```suggestion
- Non-const literal values are only supported with `ElementsClassInspector` and not reflection.
```

This is incorrect

**Review by ZacSweers — COMMENTED** [review]
Noticed some stale docs

File: docs/interop-kotlinx-metadata.md [hunk 4]
```diff
@@ -35,21 +35,49 @@ println("Is class? ${kmClass.isClass}")
...
+
+The generated representations are a _best effort_ representation of the underlying source code.
+This means that synthetic elements will be excluded from generation. Kotlin-specific language
+features like lambdas or delegation will be coerced to their idiomatic source form.
+
+To aid with this, `toTypeSpec()` and `toFileSpec()` accept optional `ClassInspector` instances
+to assist in parsing/understanding the underlying JVM code. This is important for things like
+annotations, companion objects, certain JVM modifiers, overrides, and more. While it is optional,
+represented sources can be incomplete without this information available. Reflective and javax
+`Elements` implementations are available under the `kotlinpoet-classinspector-*` artifacts.
```
Comment (Egorand) [inline docs/interop-kotlinx-metadata.md:68]: All class inspectors are in this artifact now, need to reword this statement?

File: interop/kotlinx-metadata/build.gradle.kts [hunk 5]
```diff
@@ -16,14 +16,24 @@
...
+
+tasks.compileTestKotlin {
+  kotlinOptions {
+    freeCompilerArgs = listOf("-Xjvm-default=all")
   }
 }
 
 dependencies {
+  api(libs.autoCommon)
+  api(libs.guava)
```
Comment (Egorand) [inline interop/kotlinx-metadata/build.gradle.kts:31]: Any way we can get guava off the classpath?

**Review by Egorand — COMMENTED** [review]
Nice!

**Review by Egorand — APPROVED** [review]

File: interop/kotlinx-metadata/build.gradle.kts (hunk 5 above)
Comment (ZacSweers) [reply interop/kotlinx-metadata/build.gradle.kts]: Not easily but we can at least make them implementation deps

image

File: docs/interop-kotlinx-metadata.md (hunk 4 above)
Comment (ZacSweers) [reply docs/interop-kotlinx-metadata.md]: f4b001b

File: interop/kotlinx-metadata/build.gradle.kts (hunk 5 above)
Comment (ZacSweers) [reply interop/kotlinx-metadata/build.gradle.kts]: f4b001b

Comment (Egorand) [reply interop/kotlinx-metadata/build.gradle.kts]: Yeah, that's better, assuming the types don't appear in the ABI

**Pull request merged**